server-spi-private/src/main/java/org/keycloak/authorization/policy/evaluation/DecisionPermissionCollector.java

```diff
@@ -17,141 +17,163 @@
+        // for each result entry:
+        Stream<Result> typePermissions = results.values().stream()
+            // only pass policies that have a type on them and no scopes or resources (we'll pick those up in the other streams)
+            .filter(result -> result.getPolicy() != null)
+            .filter(result -> !StringUtil.isNullOrEmpty(result.getPolicy().getType()))
+            .filter(result -> CollectionUtil.isEmpty(result.getPolicy().getResources()))
+            .filter(result -> CollectionUtil.isEmpty(result.getPolicy().getScopes()));
+        Stream<Result> resourceScopePermissions = results.values().stream()
+            // only pass policies that match the currently requested resource and either have no scopes or at least one scope matches a requested scope
+            .filter(result -> result.getPolicy() != null)
+            .filter(result -> permission.getResource() != null)
+            .filter(result -> result.getPolicy().getResources().contains(permission.getResource()))
+            .filter(result -> CollectionUtil.isEmpty(result.getPolicy().getScopes()) || result.getPolicy().getScopes().stream().anyMatch(permission.getScopes()::contains));
+        Stream<Result> scopeOnlyPermissions = results.values().stream()
+            // only pass policies that match at least one of the requested scopes and have no resources
+            .filter(result -> result.getPolicy() != null)
+            .filter(result -> CollectionUtil.isNotEmpty(permission.getScopes()))
+            .filter(result -> CollectionUtil.isEmpty(result.getPolicy().getResources()))
+            .filter(result -> result.getPolicy().getScopes().stream().anyMatch(permission.getScopes()::contains));
+        Stream<Result> grantResults = results.values().stream()
+            // make sure we don't filter out any outright grants in the process of pruning our results
+            .filter(result -> result.getPolicy() == null);
+        // remove all duplicate entries from both the original result set (because of overlapping queries) and the multiple combined streams
+        List<Result> prunedResults = Stream.of(typePermissions, resourceScopePermissions, scopeOnlyPermissions, grantResults).flatMap(resultStream -> resultStream).distinct().collect(Collectors.toList());
+
+        // if we filter out all of the fluff policies and there's nothing here then we can just return now. When policies are filtered at the DB query level
+        // we won't end up in this section of code unless a policy would pass these filters
+        if(prunedResults.isEmpty()) {
+            return;
+        }
+
+        // once queries are fixed, we can simply sort the result set. For now we have to sort the pruned results
+        List<Result> sortedResults = prunedResults.stream().sorted(Comparator.comparing(Result::getPriority)).collect(Collectors.toList());
```

Comment:
Java uses quicksort with double pivot algorithm by default. Should be able to hand around 30 million policies in less than a second in the average case on a 500Mhz CPU. 

Worst case scenario goes down to something like 20k policies (O(N^2)) but that scenario only occurs when the list is pre-sorted in reverse order. 

https://www.wolframalpha.com/input?i=+%28x+*+ln%28x%29%29%2Ff+%3D+1+where+f+%3D+500*10%5E6